engine/access/ingestion2/forest_manager.go

```diff
@@ -0,0 +1,432 @@
+
+	return true
+}
+
+// canProcessContainer determines if a container can be processed.
+//
+// Parameters:
+//   - container: the container to check
+//   - parentContainer: the parent container of the container to check
+//
+// Returns:
+//   - bool: true if the container can be processed, false otherwise
+func (fm *ForestManager) canProcessContainer(container, parentContainer *ExecutionResultContainer) bool {
+	lg := fm.log.With().
+		Hex("result_id", logging.ID(container.resultID)).
+		Logger()
+
+	// 1. Container must be pending
+	if container.Pipeline().GetState() != optimistic_sync.StatePending {
+		lg.Debug().
+			Str("state", container.Pipeline().GetState().String()).
+			Msg("container not in pending state. skipping")
+		return false
+	}
+
+	// 2. Parent must not be in pending or abandoned state
+	parentState := parentContainer.Pipeline().GetState()
+	if parentState == optimistic_sync.StatePending || parentState == optimistic_sync.StateAbandoned {
+		lg.Debug().
+			Hex("parent_id", logging.ID(parentContainer.resultID)).
+			Str("parent_state", parentState.String()).
+			Msg("parent not ready for processing")
+		return false
+	}
+
+	// 3. The result's block must have the required status
+	blockStatus, processable := fm.checkBlockStatus(container.blockHeader.Height)
+	if !processable {
+		lg.Debug().
+			Uint64("block_height", container.blockHeader.Height).
+			Str("block_status", blockStatus.String()).
+			Msg("block is not processable yet")
+		return false
+	}
+
+	// we do not need to check executors for sealed blocks
+	if blockStatus == BlockStatusSealed {
+		return true
+	}
+
+	// 4. There must be enough agreeing executors
+	if uint(container.Size()) < fm.config.RequiredAgreeingExecutors {
+		lg.Debug().
+			Uint("receipt_count", container.Size()).
+			Uint("required_count", fm.config.RequiredAgreeingExecutors).
+			Msg("insufficient agreeing executors")
+		return false
+	}
+
+	// 5. The result must have at least one of the configured required executors
+	if len(fm.config.RequiredExecutors) > 0 {
+		hasRequiredExecutor := false
+		for _, receipt := range container.receipts {
+			if _, ok := fm.config.RequiredExecutors[receipt.ExecutorID]; ok {
+				hasRequiredExecutor = true
+				break
+			}
+		}
+		if !hasRequiredExecutor {
+			lg.Debug().Msg("container missing receipts from required executors")
+			return false
+		}
+	}
+
+	return true
+}
+
+// checkBlockStatus checks if a block is processable based on the block status and the required block status.
+//
+// Parameters:
+//   - height: the height to check
+//
+// Returns:
+//   - BlockStatus: the block status
+//   - bool: true if the block meets the required block status, false otherwise
+func (fm *ForestManager) checkBlockStatus(height uint64) (BlockStatus, bool) {
+	fm.mu.RLock()
+	defer fm.mu.RUnlock()
+
+	if height > fm.finalizedHeight {
+		return BlockStatusCertified, fm.config.RequiredBlockStatus == BlockStatusCertified
+	}
+
+	if height > fm.sealedHeight {
```

Comment:
this is not sufficient because there could be results for unfinalized blocks in the forest, which could have conflicting heights.

Instead, this also needs to lookup the block by height and ensure that the returned block ID matches the results' block ID. If they match, then the block is definitely finalized, and this height check is then OK.